### PR TITLE
tests for the LAG feature

### PR DESCRIPTION
**Pull request by andriidovhan**

### CODE REVIEW

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+                        description      : "non-existent LAG port",
+                        swIdForRequest   : { NON_EXISTENT_SWITCH_ID },
+                        logicalPortNumber: 2001, // lagOffset + 1
+                        errorMsg         : "LAG port %d on switch %s not found"
+                ]
+        ]
+    }
+
+
+    //singleSw, singlePortSingleSw on a LAG port ? not sure that need to test
```
Comment (rtretyak) [inline src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy:482]: Should be tested at least manually. Single port traffic will probably have to be tested 100% manually, I doubt we can analyze that via traffgen service

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+        def lagPortIsDeleted = true
+
+        and: "LAG port is really deleted from switch"
+        !grpc.getSwitchLogicalPorts(swAddress).find { it.logicalPortNumber == lagPort }
+
+        cleanup:
+        lagPort && !lagPortIsDeleted && northboundV2.deleteLagLogicalPort(sw.dpId, lagPort)
+
+        where:
+        sw << getTopology().getActiveSwitches().unique { it.nbFormat().hardware + it.nbFormat().software }
```
[the remark this answers is not in the document]
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy]: done

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+        }
+
+        cleanup:
+        blockData && !swIsActivated && switchHelper.reviveSwitch(sw, blockData, true)
+        lagPort && northboundV2.deleteLagLogicalPort(sw.dpId, lagPort)
+    }
+
+    @Tidy
+    @Ignore("issue with traffgen")
+    def "LAG port is not deleted after switch reconnecting2"() {
```
[the remark this answers is not in the document]
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy]: deleted

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+        where:
+        data << [
+                [
+                        description: "occupied by server42",
+                        portNumber : { Switch s -> s.prop.server42Port },
+                        errorMsg   : "Physical port number %d on switch [] is server42 port."
+                ],
+                [
+                        description: "occupied by isl",
+                        portNumber : { Switch s ->
```
[the remark this answers is not in the document]
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy:349]: thx, changed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+        where:
+        data << [
+                [
+                        description      : "non-existent LAG port",
+                        swIdForRequest   : { getTopology().getActiveSwitches().first().dpId },
+                        logicalPortNumber: 1999, // lagOffset - 1
+                        errorMsg         : "LAG port %d on switch %s not found"
+                ],
+                [
+                        description      : "non-existent LAG port",
```
[the remark this answers is not in the document]
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy]: fixed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+                        logicalPortNumber: 2001, // lagOffset + 1
+                        errorMsg         : "LAG port %d on switch %s not found"
+                ]
+        ]
+    }
+
+
+    //singleSw, singlePortSingleSw on a LAG port ? not sure that need to test
+    //unable to create flow on a port which is in lag group
+    //it works in single/multi tables mode
```
[the remark this answers is not in the document]
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy]: agree

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+    //get lag port via getSwPorts  ? wait answer
+    //cant create LAG using port which is occupied by flow/mirrorPoint/
+    //getAllowedPortsForSwitch exclude ports which is > lagOffset
+    //lag portHistory - separate task
+    //try to create bfd logical port via grpc with with more than 1999
```
[the remark this answers is not in the document]
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy]: agree

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+                [
+                        description: "is more than lagOffset",
+                        portNumber : { 2008 },
+                        errorMsg   : "Physical port number %d can't be greater than LAG port offset $lagOffset."
+                ]
+        ]
+    }
+
+    @Tidy
+    def "Unable to create two LAG ports with the same physical ports inside at the same time"() {
```
[the remark this answers is not in the document]
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy]: done

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+        def createResponse = northboundV2.createLagLogicalPort(sw.dpId, payload)
+
+        then: "Response reports successful creation of the LAG port"
+        with(createResponse) {
+            logicalPortNumber > 0
+            portNumbers.sort() == portsArray.sort()
+        }
+        def lagPort = createResponse.logicalPortNumber
+
+        and: "Flow is really created"
```
[the remark this answers is not in the document]
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy:67]: I didn't get it.

[the remark this answers is not in the document]
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy]: fixed

File: src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy (hunk 1 above)
Comment (andriidovhan) [reply src-java/testing/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/switches/LagPortSpec.groovy:486]: test was added

**Pull request merged**